distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+        peer_addr = self.prefix + address
+        return self.comm_class(protocol, local_addr, peer_addr, deserialize=deserialize)
+
+    def _get_extra_kwargs(self, address, **kwargs):
+        _error_if_require_encryption(address, **kwargs)
+        return {}
+
+
+class TLSConnector(TCPConnector):
+    prefix = "tls://"
+    comm_class = TLS
+
+    def _get_extra_kwargs(self, address, **kwargs):
+        ctx = _expect_tls_context(kwargs)
+        return {"ssl": ctx}
+
+
+class TCPListener(Listener):
+    prefix = "tcp://"
+    comm_class = TCP
+
+    def __init__(
+        self,
+        address,
+        comm_handler,
+        deserialize=True,
+        allow_offload=True,
+        default_host=None,
+        default_port=0,
+        **kwargs,
+    ):
+        self.ip, self.port = parse_host_port(address, default_port)
+        self.default_host = default_host
+        self.comm_handler = comm_handler
+        self.deserialize = deserialize
+        self.allow_offload = allow_offload
+        self._extra_kwargs = self._get_extra_kwargs(address, **kwargs)
+        self.bound_address = None
+
+    def _get_extra_kwargs(self, address, **kwargs):
+        _error_if_require_encryption(address, **kwargs)
+        return {}
+
+    def _on_connection(self, protocol):
+        comm = self.comm_class(
+            protocol,
+            local_addr=self.prefix + protocol.local_addr,
+            peer_addr=self.prefix + protocol.peer_addr,
+            deserialize=self.deserialize,
+        )
+        comm.allow_offload = self.allow_offload
+        asyncio.ensure_future(self._comm_handler(comm))
+
+    async def _comm_handler(self, comm):
+        try:
+            await self.on_connection(comm)
+        except CommClosedError:
+            logger.debug("Connection closed before handshake completed")
+            return
+        await self.comm_handler(comm)
+
+    async def _start_all_interfaces_with_random_port(self):
+        """Due to a design decision in asyncio, listening on `("", 0)` will
+        result in two different random ports being used (one for IPV4, one for
+        IPV6), rather than both interfaces sharing the same random port. We
+        work around this here. See https://bugs.python.org/issue45693 for more
+        info."""
+        loop = asyncio.get_running_loop()
+        # Typically resolves to list with length == 2 (one IPV4, one IPV6).
+        infos = await loop.getaddrinfo(
+            None,
+            0,
+            family=socket.AF_UNSPEC,
+            type=socket.SOCK_STREAM,
+            flags=socket.AI_PASSIVE,
+            proto=0,
+        )
+        # Sort infos to always bind ipv4 before ipv6
+        infos = sorted(infos, key=lambda x: x[0].name)
+        # This code is a simplified and modified version of that found in
+        # cpython here:
+        # https://github.com/python/cpython/blob/401272e6e660445d6556d5cd4db88ed4267a50b3/Lib/asyncio/base_events.py#L1439
```

Comment:
How different is this? Should we include the license file? Also what was the reason for needing to copy and modify?

Reply:
We already copy bits of cpython out elsewhere, this has been sufficient so far.

> Also what was the reason for needing to copy and modify?

See https://github.com/dask/distributed/blob/8351fbfffdd77fbebed257fe07ca5daebbec30bd/distributed/comm/asyncio_tcp.py#L587-L591.